pdr_backend/sim/sim_engine.py

```diff
@@ -261,3 +291,113 @@ def save_state(self, i: int, N: int):
             return False, False
 
         return True, False
+
+    @enforce_types
+    def _get_predictions_signals_data(
+        self, start_slot: int, end_slot: int
+    ) -> Dict[int, Optional[float]]:
+        result_dict: Dict[int, Optional[float]] = {}
+        db = DuckDBDataStore(self.ppss.lake_ss.lake_dir)
+        query_cont = f"""
+            SELECT
+                contract
+            FROM
+                pdr_predictions
+            WHERE
+                timeframe = '{self.predict_feed.timeframe}'
+                AND pair = '{self.predict_feed.pair.pair_str}'
+            LIMIT 1
+        """
+        feed_contract_addr = db.query_data(query_cont)
+        if len(feed_contract_addr) == 0:
+            logger.error("Contract address for the given feed not found in database")
+            return result_dict
+
+        query = f"""
+            SELECT
+                slot,
+                CASE
+                    WHEN roundSumStakes = 0.0 THEN NULL
+                    ELSE roundSumStakesUp / roundSumStakes
+                END AS probUp
+            FROM
+                pdr_payouts
+            WHERE
+                slot > {start_slot}
+                AND slot < {end_slot}
+                AND ID LIKE '{feed_contract_addr.row(0)[0]}%'
+        """
+
+        df: pl.DataFrame = db.query_data(query)
+
+        for i in range(len(df)):
```

Comment:
isn't there a better iterator? The presence of a `for... i` can sometimes be a code smell

Reply:
I can improve it to this:
`for row in df.to_dicts():
            if row["probUp"] is not None:
                result_dict[row["slot"]] = row["probUp"]`